### ISSUE TITLE
CMakeLists.txt build type

### ISSUE THREAD

**Issue by rtlprmft** [imbs-hl/ranger#447]
I think the correct way to set the default build type for cmake is
SET(CMAKE_BUILD_TYPE_INIT Release)
and not 
if(NOT CMAKE_BUILD_TYPE)
  set(CMAKE_BUILD_TYPE Release)
endif()

Although both seem to work the same, way the latter only work on following commands and doesn't show up in the cmake cache which is confusing.

**Comment by rtlprmft** [comment]
Properly setting up pthread for all compilers should work with
FIND_PACKAGE(Threads  REQUIRED) 
If the thread library is required add Threads::Threads as library

**Comment by mnwright** [comment]
I cannot find a proper documentation of `CMAKE_BUILD_TYPE_INIT`. Do you have a link? 

Your suggestion for pthread looks good. According to [this SO answer](https://stackoverflow.com/a/29871891) it requires CMake 3.1., which might be a problem for some people (it seems many people use ranger on clusters with very old software).

**Comment by rtlprmft** [comment]
I think it is not well documented because "it was originally considered internal".

I have just tested: SET(CMAKE_BUILD_TYPE_INIT Release) at the beginning of CMakeLists without the 
if(NOT CMAKE_BUILD_TYPE)  
set(CMAKE_BUILD_TYPE Release) 
endif() 
makes the compiler use Release definitions (-O3 -NDEBUG)

No SET(CMAKE_BUILD_TYPE_INIT Release)  but your IF()/ENDIF() uses no additional flags (i.e. no compiler optimization)

In both cases "ccmake ." displays an empty string in CMAKE_BUILD_TYPE.

A third solution is to do
if(NOT CMAKE_BUILD_TYPE)
SET(CMAKE_BUILD_TYPE Release CACHE STRING "Build Configuration" FORCE)
endif()
This works as expected and makes CMAKE_BUILD_TYPE show up in the cache entries ("ccmake .") as well.

The reason why I personally prefer the _INIT option is that when you look at the explanation for the cache Entry (displayed by ccmake .) with the _INIT option, the original text is displayed (which is quite complete and lists all options) and in the later case only the "Build Configuration" is displayed without any hint what the options are. (Funny side note: until you change that option with CMAKE, then the original text is displayed again)

Just out of curiosity. Was there a reason not to use SET(CMAKE_CXX_STANDARD 11)?

**Comment by rtlprmft** [comment]
Btw, if that helps: I run cmake 3.10, but recently tested our build systems also on 3.1 and 3.15.

**Comment by mnwright** [comment]
Hm... I also just tested. I created a fresh build directory and `make`, `make debug` and `make release`, each with `VERBOSE=1`. 

With the current approach (if ... ): 
- just make: -O3 and -DNDEBUG flags
- make debug: -g flag
- make release: -O3 and -DNDEBUG flags

With your suggestion (init): 
- just make: no flags
- make debug: -g flag
- make release: -O3 and -DNDEBUG flags

I'm using cmake 3.8.2 on a Mac. 

> Just out of curiosity. Was there a reason not to use SET(CMAKE_CXX_STANDARD 11)?

I think this was also introduced in CMake 3.1.0, which was released after we created the CMakeLists.txt.
